File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/services/RefundHandling.java
package it.giannibombelli.wsc2026.payment.application.services;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.events.EventPublisher;
import it.giannibombelli.wsc2026.common.application.events.EventSubscriber;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.payment.application.commands.RefundTransaction;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.RefundRequested;
import it.giannibombelli.wsc2026.payment.domain.events.RefundResultEvents;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.Transaction;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionStatus;
import it.giannibombelli.wsc2026.payment.domain.policies.TransactionRefund;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProvider;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProviderResult;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static java.util.Objects.requireNonNull;

public final class RefundHandling implements EventSubscriber<RefundRequested> {
    private final PaymentRepository paymentRepository;
    private final Map<String, PaymentProvider> providers;
    private final TransactionRefund transactionRefund;
    private final EventPublisher<PaymentEvent> eventPublisher;

    public RefundHandling(PaymentRepository paymentRepository,
                          Map<String, PaymentProvider> providers,
                          TransactionRefund transactionRefund,
                          EventPublisher<PaymentEvent> eventPublisher) {
        this.paymentRepository = requireNonNull(paymentRepository);
        this.providers = requireNonNull(providers);
        this.transactionRefund = requireNonNull(transactionRefund);
        this.eventPublisher = requireNonNull(eventPublisher);
    }

    @Override
    public void on(RefundRequested event) {
        Require.requireArgument(event, "event");

        RefundTransaction cmd = transactionRefund.evaluate(event);

        Payment payment = paymentRepository.findById(cmd.aggregateId())
            .orElseThrow(() -> new IllegalArgumentException("payment not found"));

        List<Transaction> acceptedTransactions = payment.transactions().stream()
            .filter(t -> t.status() == TransactionStatus.ACCEPTED)
            .toList();

        if (acceptedTransactions.isEmpty()) {
            throw new IllegalStateException("payment has no accepted transaction");
        }

        Money refundedSum = Money.zero();
        Transaction failedTransaction = null;
        String failureReason = null;

        for (Transaction transaction : acceptedTransactions) {
            PaymentProvider provider = providers.get(transaction.provider().label());
            if (provider == null) {
                failedTransaction = transaction;
                failureReason = "unknown provider: " + transaction.provider().label();
                break;
            }

            PaymentProviderResult result = provider.refund(
                payment.id().value(),
                transaction.providerReference().value(),
                transaction.amount()
            );

            switch (result) {
                case PaymentProviderResult.Success s -> {
                    payment.markTransactionRefunded(transaction.id());
                    refundedSum = refundedSum.plus(transaction.amount());
                }
                case PaymentProviderResult.Failure f -> {
                    failedTransaction = transaction;
                    failureReason = f.reason().value();
                }
            }

            if (failedTransaction != null) {
                break;
            }
        }

        if (failedTransaction != null) {
            RefundResultEvents.TransactionNotRefunded notRefunded = payment.rejectRefund(
                failedTransaction.provider(), failedTransaction.providerReference(), new Description(failureReason)
            );
            paymentRepository.save(payment);
            eventPublisher.publish(notRefunded);
            return;
        }

        RefundTransaction refundedCmd = new RefundTransaction(payment.id(), refundedSum);
        RefundResultEvents.TransactionRefunded refunded = payment.refundTransaction(refundedCmd.amount());
        paymentRepository.save(payment);
        eventPublisher.publish(refunded);
    }
}

File: typescript/src/payment/application/services/refundHandling.ts
// Servizio applicativo che gestisce le richieste di rimborso contattando il provider.

import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import type { EventPublisher } from '@/common/application/events/eventPublisher.js';
import type { EventSubscriber } from '@/common/application/events/eventSubscriber.js';
import { PaymentEvent } from '../../domain/events/paymentEvent.js';
import { RefundRequested } from '../../domain/events/refundRequested.js';
import type {
  TransactionRefunded,
  TransactionNotRefunded,
} from '../../domain/events/refundResultEvents.js';
import { PaymentNotFoundException } from '../../domain/payment/paymentNotFoundException.js';
import { Transaction } from '../../domain/payment/transaction.js';
import { TransactionStatus } from '../../domain/payment/transactionStatus.js';
import { TransactionRefund } from '../../application/policies/transactionRefund.js';
import type { PaymentProvider } from '../../domain/ports/paymentProvider.js';
import { PaymentProviderResult } from '../../domain/ports/paymentProviderResult.js';
import type { PaymentRepository } from '../../domain/ports/paymentRepository.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class RefundHandling implements EventSubscriber<RefundRequested> {
  private readonly _paymentRepository: PaymentRepository;
  private readonly _providers: Record<string, PaymentProvider>;
  private readonly _transactionRefund: TransactionRefund;
  private readonly _eventPublisher: EventPublisher<PaymentEvent>;

  constructor(
    paymentRepository: PaymentRepository,
    providers: Record<string, PaymentProvider>,
    transactionRefund: TransactionRefund,
    eventPublisher: EventPublisher<PaymentEvent>
  ) {
    requireDependency(paymentRepository, "paymentRepository");
    requireDependency(providers, "providers");
    requireDependency(transactionRefund, "transactionRefund");
    requireDependency(eventPublisher, "eventPublisher");
    this._paymentRepository = paymentRepository;
    this._providers = providers;
    this._transactionRefund = transactionRefund;
    this._eventPublisher = eventPublisher;
  }

  on(event: RefundRequested): void {
    requireArgument(event, 'event');

    const cmd = this._transactionRefund.evaluate(event);

    const payment = this._paymentRepository.findById(cmd.aggregateId);
    if (payment === null) throw new PaymentNotFoundException();

    const acceptedTransactions = payment.transactions().filter((t) => t.status() === TransactionStatus.ACCEPTED);
    if (acceptedTransactions.length === 0) {
      throw new Error('payment has no accepted transaction');
    }

    let refundedSum = Money.zero();
    let failedTransaction: Transaction | null = null;
    let failureReason: Description | null = null;

    for (const transaction of acceptedTransactions) {
      const provider = this._providers[transaction.provider()];
      if (provider === undefined) {
        failedTransaction = transaction;
        failureReason = new Description(`unknown provider: ${transaction.provider()}`);
        break;
      }

      const result: PaymentProviderResult = provider.refund(
        payment.id().value,
        transaction.providerReference()!.value,
        transaction.amount()
      );

      if (result.kind === 'success') {
        payment.markTransactionRefunded(transaction.id());
        refundedSum = refundedSum.plus(transaction.amount());
      } else {
        failedTransaction = transaction;
        failureReason = result.reason;
      }

      if (failedTransaction !== null) {
        break;
      }
    }

    if (failedTransaction !== null) {
      const notRefunded: TransactionNotRefunded = payment.rejectRefund(
        failedTransaction.provider(),
        failedTransaction.providerReference()!,
        failureReason!
      );
      this._paymentRepository.save(payment);
      this._eventPublisher.publish(notRefunded);
      return;
    }

    const refunded: TransactionRefunded = payment.refundTransaction(refundedSum);
    this._paymentRepository.save(payment);
    this._eventPublisher.publish(refunded);
  }
}
